File: windows/tapenexus/views/stats_view.py
"""Insights tab: a read-only dashboard built from the Library archive and the
achievements stats — downloads over time, top hosts, formats used, and the
cumulative totals already tracked for the leaderboard. Pure UI, no backend."""
from __future__ import annotations

from collections import Counter
from datetime import date, datetime, timedelta
from typing import List, Tuple

from PySide6.QtCore import Qt
from PySide6.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QLabel, QScrollArea, QFrame,
    QSizePolicy,
)

from . import theme


def _week_start(d: date) -> date:
    """Monday of the week containing d."""
    return d - timedelta(days=d.weekday())


def _fmt_bytes(n: int) -> str:
    if n <= 0:
        return "0 B"
    units = ["B", "KB", "MB", "GB", "TB"]
    f = float(n)
    i = 0
    while f >= 1024 and i < len(units) - 1:
        f /= 1024.0
        i += 1
    return f"{int(f)} {units[i]}" if i <= 1 else f"{f:.1f} {units[i]}"


class _StatCard(QFrame):
    def __init__(self, icon: str, value: str, label: str, tint: str) -> None:
        super().__init__()
        self.setStyleSheet(
            f"QFrame {{ background: {theme.PANEL}; border: 1px solid {theme.LINE};"
            f" border-radius: 12px; }}")
        v = QVBoxLayout(self)
        v.setContentsMargins(12, 12, 12, 12)
        v.setSpacing(6)
        ic = QLabel(icon)
        ic.setStyleSheet(f"font-size: 16px; color: {tint};")
        v.addWidget(ic)
        val = QLabel(value)
        val.setStyleSheet(f"font-size: 18px; font-weight: 700; color: {theme.TEXT};")
        v.addWidget(val)
        lbl = QLabel(label)
        lbl.setStyleSheet(f"font-size: 10px; color: {theme.MUTED};")
        v.addWidget(lbl)


class _HBar(QFrame):
    """One horizontal bar row: label | track+fill | count."""

    def __init__(self, label: str, count: int, max_count: int, tint: str) -> None:
        super().__init__()
        self.setStyleSheet("background: transparent;")
        row = QHBoxLayout(self)
        row.setContentsMargins(0, 0, 0, 0)
        row.setSpacing(10)
        lbl = QLabel(label)
        lbl.setStyleSheet(f"font-size: 11px; font-weight: 500; color: {theme.TEXT};")
        lbl.setFixedWidth(170)
        row.addWidget(lbl)
        track = QFrame()
        track.setFixedHeight(12)
        track.setStyleSheet(f"background: {theme.PANEL2}; border-radius: 4px;")
        fill = QFrame(track)
        fill.setFixedHeight(12)
        frac = (count / max(max_count, 1)) if max_count > 0 else 0.0
        # Track width is unknown until laid out; use a fixed 260px track so the
        # fill width is deterministic without a resize event.
        track.setFixedWidth(260)
        fill.setFixedWidth(max(1, int(260 * frac)))
        fill.setStyleSheet(f"background: {tint}; border-radius: 4px;")
        row.addWidget(track)
        cnt = QLabel(str(count))
        cnt.setStyleSheet(f"font-size: 11px; font-weight: 600; color: {theme.MUTED};")
        cnt.setFixedWidth(34)
        cnt.setAlignment(Qt.AlignRight | Qt.AlignVCenter)
        row.addWidget(cnt)


def _section(title: str, hint: str = "") -> QFrame:
    card = QFrame()
    card.setStyleSheet(
        f"QFrame {{ background: {theme.PANEL}; border: 1px solid {theme.LINE};"
        f" border-radius: 12px; }}")
    v = QVBoxLayout(card)
    v.setContentsMargins(14, 14, 14, 14)
    v.setSpacing(10)
    head = QHBoxLayout()
    t = QLabel(title)
    t.setStyleSheet(f"font-size: 13px; font-weight: 600; color: {theme.TEXT};")
    head.addWidget(t)
    head.addStretch(1)
    if hint:
        h = QLabel(hint)
        h.setStyleSheet(f"font-size: 10px; color: {theme.MUTED};")
        head.addWidget(h)
    v.addLayout(head)
    return card, v


class StatsView(QWidget):
    def __init__(self, state) -> None:
        super().__init__()
        self.state = state

        outer = QVBoxLayout(self)
        outer.setContentsMargins(16, 12, 16, 12)
        outer.setSpacing(12)

        title = QLabel("Insights")
        title.setStyleSheet(f"font-size: 22px; font-weight: 700; color: {theme.TEXT};")
        outer.addWidget(title)
        sub = QLabel("Your download activity at a glance.")
        sub.setStyleSheet(f"font-size: 12px; color: {theme.MUTED};")
        outer.addWidget(sub)

        # Scrollable body so the dashboard never clips on small windows.
        scroll = QScrollArea()
        scroll.setWidgetResizable(True)
        scroll.setFrameShape(QFrame.NoFrame)
        scroll.setStyleSheet(
            f"QScrollArea {{ background: transparent; border: none; }}"
            f" QScrollBar:vertical {{ background: {theme.PANEL}; width: 8px; }}")
        body = QWidget()
        body.setStyleSheet("background: transparent;")
        self.body_layout = QVBoxLayout(body)
        self.body_layout.setContentsMargins(0, 0, 0, 0)
        self.body_layout.setSpacing(14)
        self.body_layout.addStretch(1)
        scroll.setWidget(body)
        outer.addWidget(scroll, 1)

        # Refresh whenever the library or achievements change.
        try:
            self.state.library.library_changed.connect(self._rebuild)
        except Exception:
            pass
        self._rebuild()

    # ── build / rebuild ──────────────────────────────────────────────────────
    def _clear(self) -> None:
        while self.body_layout.count() > 1:
            it = self.body_layout.takeAt(0)
            w = it.widget()
            if w is not None:
                w.deleteLater()

    def _rebuild(self) -> None:
        self._clear()
        stats = self.state.achievements.stats
        entries = list(self.state.library.entries)

        # Stat cards row.
        cards = QHBoxLayout()
        cards.setSpacing(10)
        cards.addWidget(_StatCard("⬇️", str(stats.total_completed), "Downloads", theme.ACCENT))
        cards.addWidget(_StatCard("💾", self.state.achievements.formatted_total_bytes(),
                                  "Total data", theme.ACCENT2))
        cards.addWidget(_StatCard("🔥", f"{stats.best_streak()}d", "Best streak", theme.WARN))
        cards.addWidget(_StatCard("🌍", str(len(stats.hosts_seen)), "Hosts", theme.BLUE))
        cards.addWidget(_StatCard("🏆", str(stats.score()), "Score", theme.OK))
        cards.addStretch(1)
        host = QWidget()
        host.setStyleSheet("background: transparent;")
        host.setLayout(cards)
        self._insert(host)

        self._insert(self._weekly_card(entries))
        self._insert(self._top_card("Top hosts", f"{len(stats.hosts_seen)} distinct",
                                    self._top_counts([e.host() for e in entries], 6),
                                    theme.ACCENT))
        self._insert(self._top_card("Formats used", "",
                                    self._top_counts(
                                        [e.format_desc or "unknown" for e in entries], 6),
                                    theme.ACCENT2))

    def _insert(self, w: QWidget) -> None:
        self.body_layout.insertWidget(self.body_layout.count() - 1, w)

    # ── charts ───────────────────────────────────────────────────────────────
    def _weekly_card(self, entries) -> QFrame:
        buckets = self._weekly_buckets(entries)
        max_count = max((b[1] for b in buckets), default=0) or 1
        card, v = _section("Downloads over the last 12 weeks",
                           hint=f"{len(entries)} archived")
        row = QHBoxLayout()
        row.setSpacing(6)
        for label, count in buckets:
            col = QVBoxLayout()
            col.setSpacing(4)
            col.setAlignment(Qt.AlignBottom)
            n = QLabel(str(count) if count > 0 else "")
            n.setStyleSheet(f"font-size: 9px; font-weight: 600; color: {theme.MUTED};")
            n.setAlignment(Qt.AlignCenter)
            col.addWidget(n)
            bar = QFrame()
            h = int(90 * count / max_count) if count > 0 else 0
            bar.setFixedHeight(max(3 if count > 0 else 0, h))
            bar.setFixedWidth(18)
            bar.setStyleSheet(
                f"background: {theme.ACCENT}; border-radius: 3px;")
            col.addWidget(bar)
            lb = QLabel(label)
            lb.setStyleSheet(f"font-size: 8px; color: {theme.MUTED};")
            lb.setAlignment(Qt.AlignCenter)
            col.addWidget(lb)
            wrap = QWidget()
            wrap.setStyleSheet("background: transparent;")
            wrap.setLayout(col)
            wrap.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Fixed)
            row.addWidget(wrap)
        row.addStretch(1)
        v.addLayout(row)
        return card

    def _top_card(self, title: str, hint: str, rows: List[Tuple[str, int]],
                  tint: str) -> QFrame:
        card, v = _section(title, hint=hint)
        if not rows:
            empty = QLabel("Nothing archived yet — finish a download to see stats here.")
            empty.setStyleSheet(f"font-size: 11px; color: {theme.MUTED};")
            v.addWidget(empty)
        else:
            top = rows[0][1]
            for label, count in rows:
                v.addWidget(_HBar(label, count, top, tint))
        return card

    # ── data helpers ─────────────────────────────────────────────────────────
    def _weekly_buckets(self, entries) -> List[Tuple[str, int]]:
        today = date.today()
        starts = [_week_start(today - timedelta(weeks=i)) for i in range(11, -1, -1)]
        counts = {s: 0 for s in starts}
        for e in entries:
            try:
                d = datetime.fromisoformat(e.completed_at).date() if e.completed_at else None
            except Exception:
                d = None
            if d is None:
                continue
            ws = _week_start(d)
            if ws in counts:
                counts[ws] += 1
        out = []
        for s in starts:
            iso = s.isocalendar()
            out.append((f"{iso[1]}/{str(iso[0])[2:]}", counts[s]))
        return out

    def _top_counts(self, keys: List[str], limit: int) -> List[Tuple[str, int]]:
        c = Counter(keys)
        return c.most_common(limit)

File: Sources/TapeNexus/Views/StatsView.swift
import SwiftUI

/// Insights tab — a read-only dashboard built entirely from data the app
/// already collects: `LibraryStore.entries` (per-download host/format/date) and
/// `AchievementsManager.stats` (cumulative totals, streaks, score). No backend,
/// no sync changes. Charts are hand-rolled (GeometryReader + Rectangle) so we
/// don't pull in the Charts framework and bump the build.
struct StatsView: View {
    @EnvironmentObject var state: AppState

    private var entries: [LibraryEntry] { state.library.entries }
    private var stats: AchievementStats { state.achievements.stats }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 18) {
                header
                statCards
                weeklyChart
                topHosts
                topFormats
            }
            .padding(20)
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .background(Theme.bg)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
    }

    // MARK: Header

    private var header: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("Insights")
                .font(.system(size: 22, weight: .bold)).foregroundStyle(Theme.text)
            Text("Your download activity at a glance.")
                .font(.system(size: 12)).foregroundStyle(Theme.muted)
        }
    }

    // MARK: Stat cards

    private var statCards: some View {
        HStack(spacing: 10) {
            StatCard(system: "arrow.down.circle.fill", value: "\(stats.totalCompleted)",
                     label: "Downloads", tint: Theme.accent)
            StatCard(system: "externaldrive.fill", value: state.achievements.formattedTotalBytes,
                     label: "Total data", tint: Theme.accent2)
            StatCard(system: "flame.fill", value: "\(stats.bestStreak)d",
                     label: "Best streak", tint: Theme.warn)
            StatCard(system: "globe", value: "\(stats.hostsSeen.count)",
                     label: "Hosts", tint: Theme.blue)
            StatCard(system: "trophy.fill", value: "\(stats.score)",
                     label: "Score", tint: Theme.ok)
        }
    }

    // MARK: Weekly downloads chart

    private var weeklyChart: some View {
        let buckets = weeklyBuckets
        let maxCount = max(1, buckets.map(\.count).max() ?? 0)
        return card {
            VStack(alignment: .leading, spacing: 12) {
                chartTitle("Downloads over the last 12 weeks", hint: "\(entries.count) archived")
                HStack(alignment: .bottom, spacing: 6) {
                    ForEach(buckets) { b in
                        VStack(spacing: 4) {
                            Text(b.count > 0 ? "\(b.count)" : "")
                                .font(.system(size: 9, weight: .semibold))
                                .foregroundStyle(Theme.muted)
                            GeometryReader { geo in
                                let h = geo.size.height * CGFloat(b.count) / CGFloat(maxCount)
                                RoundedRectangle(cornerRadius: 3)
                                    .fill(LinearGradient(colors: [Theme.accent, Theme.accent2],
                                                         startPoint: .bottom, endPoint: .top))
                                    .frame(height: max(b.count > 0 ? 3 : 0, h))
                                    .frame(maxHeight: .infinity, alignment: .bottom)
                            }
                            .frame(width: 18, height: 90)
                            Text(b.label)
                                .font(.system(size: 8)).foregroundStyle(Theme.muted)
                        }
                        .frame(maxWidth: .infinity)
                    }
                }
            }
        }
    }

    // MARK: Top hosts

    private var topHosts: some View {
        let rows = topCounts(keyed: { $0.host }, limit: 6)
        return card {
            VStack(alignment: .leading, spacing: 10) {
                chartTitle("Top hosts", hint: "\(stats.hostsSeen.count) distinct")
                if rows.isEmpty { emptyHint }
                else {
                    ForEach(rows, id: \.key) { r in
                        HBarRow(label: r.key, count: r.count,
                                maxCount: rows[0].count, tint: Theme.accent)
                    }
                }
            }
        }
    }

    // MARK: Top formats

    private var topFormats: some View {
        let rows = topCounts(keyed: { $0.formatDesc.isEmpty ? "unknown" : $0.formatDesc }, limit: 6)
        return card {
            VStack(alignment: .leading, spacing: 10) {
                chartTitle("Formats used", hint: nil)
                if rows.isEmpty { emptyHint }
                else {
                    ForEach(rows, id: \.key) { r in
                        HBarRow(label: r.key, count: r.count,
                                maxCount: rows[0].count, tint: Theme.accent2)
                    }
                }
            }
        }
    }

    // MARK: Helpers

    private var emptyHint: some View {
        Text("Nothing archived yet — finish a download to see stats here.")
            .font(.system(size: 11)).foregroundStyle(Theme.muted)
    }

    @ViewBuilder
    private func card<Content: View>(@ViewBuilder _ content: () -> Content) -> some View {
        VStack(alignment: .leading, spacing: 0) { content() }
            .padding(14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Theme.panel)
            .overlay(RoundedRectangle(cornerRadius: 12).stroke(Theme.line))
            .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    private func chartTitle(_ title: String, hint: String?) -> some View {
        HStack {
            Text(title).font(.system(size: 13, weight: .semibold)).foregroundStyle(Theme.text)
            Spacer()
            if let hint { Text(hint).font(.system(size: 10)).foregroundStyle(Theme.muted) }
        }
    }

    /// Bucket archived entries into the last 12 calendar weeks (oldest → newest).
    private var weeklyBuckets: [WeekBucket] {
        let cal = Calendar.current
        let now = Date()
        guard let thisWeek = cal.dateInterval(of: .weekOfYear, for: now) else { return [] }
        var starts: [Date] = []
        for i in stride(from: 11, through: 0, by: -1) {
            if let s = cal.date(byAdding: .weekOfYear, value: -i, to: thisWeek.start) {
                starts.append(s)
            }
        }
        var counts = Dictionary(uniqueKeysWithValues: starts.map { ($0, 0) })
        for e in entries {
            guard let wk = cal.dateInterval(of: .weekOfYear, for: e.completedAt)?.start,
                  counts[wk] != nil else { continue }
            counts[wk, default: 0] += 1
        }
        let nf = DateFormatter()
        nf.dateFormat = "w/yy"
        return starts.map { WeekBucket(label: nf.string(from: $0), count: counts[$0] ?? 0) }
    }

    /// Top-N grouped counts of archived entries by a key, sorted by count desc.
    private func topCounts(keyed: (LibraryEntry) -> String, limit: Int) -> [(key: String, count: Int)] {
        var g = [String: Int]()
        for e in entries { g[keyed(e), default: 0] += 1 }
        return g.sorted { $0.value > $1.value }.prefix(limit).map { ($0.key, $0.value) }
    }
}

// MARK: - Subviews

private struct StatCard: View {
    let system: String
    let value: String
    let label: String
    let tint: Color
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Image(systemName: system).font(.system(size: 16)).foregroundStyle(tint)
            Text(value).font(.system(size: 18, weight: .bold)).foregroundStyle(Theme.text)
                .lineLimit(1).minimumScaleFactor(0.7)
            Text(label).font(.system(size: 10)).foregroundStyle(Theme.muted)
        }
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.panel)
        .overlay(RoundedRectangle(cornerRadius: 12).stroke(Theme.line))
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }
}

private struct HBarRow: View {
    let label: String
    let count: Int
    let maxCount: Int
    let tint: Color
    var body: some View {
        HStack(spacing: 10) {
            Text(label)
                .font(.system(size: 11, weight: .medium)).foregroundStyle(Theme.text)
                .lineLimit(1).truncationMode(.tail)
                .frame(width: 170, alignment: .leading)
            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    RoundedRectangle(cornerRadius: 4).fill(Theme.panel2)
                    RoundedRectangle(cornerRadius: 4).fill(tint.opacity(0.85))
                        .frame(width: maxCount > 0
                               ? geo.size.width * CGFloat(count) / CGFloat(max(maxCount, 1)) : 0)
                }
            }
            .frame(height: 12)
            Text("\(count)")
                .font(.system(size: 11, weight: .semibold)).foregroundStyle(Theme.muted)
                .frame(width: 34, alignment: .trailing)
        }
    }
}

private struct WeekBucket: Identifiable {
    let id = UUID()
    let label: String
    let count: Int
}